Add unit test environment that runs quick tests under src/pudl

rousik opened the pull request:
this new testenv is also included in the `ci` test suite

This also necessitated fixing `excel_test.py` to account for the recent changes. Function signatures needed to be changed (year --> generic partition logic), xlsx metadata files had to be transposed and one bug in the function call `process_raw` was discovered as part of this refactoring.

rousik commented:
looks like unit tests do not like `--fast` argument. Perhaps we can simply remove it

zaneselvans commented on tox.ini:
```diff
@@ -114,6 +115,15 @@ extras =
 commands =
     pytest {posargs} --disable-warnings --log-cli-level=info --log-cli-format="%(asctime)s [%(levelname)8s] %(name)s:%(lineno)s %(message)s" --log-date-format="%Y-%m-%d %H:%M:%S" --cov-report=term --cov-report=xml --cov-config=.coveragerc --cov={envsitepackagesdir}/pudl test/etl_test.py test/glue_test.py test/fast_output_test.py test/interim_etl_test.py test/datastore_test.py
 
+
+# Runs unit tests under src/pudl
+[testenv:unit]
+extras =
+    test
+commands =
+    pytest {posargs} --disable-warnings --log-cli-level=info --log-cli-format="%(asctime)s [%(levelname)8s] %(name)s:%(lineno)s %(message)s" --log-date-format="%Y-%m-%d %H:%M:%S" --cov-report=term --cov-report=xml --cov-config=.coveragerc --cov={envsitepackagesdir}/pudl src/pudl
+
+
```
Huh, now that I see this broken out, I wonder if it doesn't make sense to break out the other individual collections of tests that we want to run, so they can be run individually easily if necessary. Is there some way we can store and re-use this long-ass pytest command? This creates a single, shared coverage report, right?

rousik replied:
Hah. I must admit that I do not have full understanding of tox/pytest and code coverage setup to be able to answer this question. I will look into it and see if I can figure this out.

rousik replied:
Actually, after reading https://docs.pytest.org/en/stable/unittest.html it seems that we can simply rely on `pytest tests` to run this. Let me adapt the code to do so.

zaneselvans commented:
By removing `--fast` do you mean just always do a single year of testing? Since that's pretty much what we do already, instead running the ETL scripts to test the integration of all the data. Especially since we want to move to running the full ETL every night anyway.

Hmm, or maybe we need to point pytest at the `pytest.ini` file that's in the `test` directory even when it's running tests in the `src` directory. Or maybe move `pytest.ini` into the root directory of the repo if we're going to have tests in both places?

rousik commented:
> By removing `--fast` do you mean just always do a single year of testing? Since that's pretty much what we do already, instead running the ETL scripts to test the integration of all the data. Especially since we want to move to running the full ETL every night anyway.

I think that `--fast` should not be relevant when it comes to unit tests. These should be synthetic and quick tests verifying that functions do what they're expected to do. In other words, I do not expect that we'd be running significant portions of ETL when running unit tests and ideally we would not depend on any external (and possibly changing) data.

> 
> Hmm, or maybe we need to point pytest at the `pytest.ini` file that's in the `test` directory even when it's running tests in the `src` directory. Or maybe move `pytest.ini` into the root directory of the repo if we're going to have tests in both places?

From my brief analysis, it seems that we could configure pytest within `tox.ini`. Right now I'm not exactly sure what kinds of settings we might want to put there.

rousik commented:
I have simplified the setup, fixed `excel_tests.py`, removed tests from documentation and source distribution and added unit tests as part of the pre-submit suite - this should help us keep those tests from accidentally breaking.

zaneselvans commented on tox.ini:
```diff
@@ -178,6 +187,7 @@ commands =
     {[testenv:docs]commands}
     {[testenv:etl]commands}
     {[testenv:scripts]commands}
+    {[testenv:unit]commands}
```
I believe these are run in order, and since the unit tests run quickly, maybe it makes sense to put them before the ETL, so that if they pick up a problem, the tests fail quickly, rather than taking half an hour.

zaneselvans reviewed (changes requested):
Let's have the unit tests run before the ETL in the CI environment, just so they fail quickly if they're going to fail.

zaneselvans commented:
I ran the `unit` tests alone with `tox -rve unit` locally and the completed quickly (6 seconds!) but it didn't look like they generated any coverage data. I know there's a way to accumulate coverage from different test environments in a single output but I haven't messed with it.

rousik commented:
Right now, `excel_test.py` is the only unit test we have. `datastore_refactoring` adds more and it has been accidentally merged into `tox-unit-tests` which is why it is run there as well.

zaneselvans commented:
@rousik can you merge in the changes from `sprint29` that came from the `datastore_refactoring` branch? There's a conflict in `tox.ini` between the `unit` and `etl` test environments. I took a quick look but I'm not sure what's supposed to be happening there in the new setup.

rousik commented:
I have merged `sprint29` back into this branch, added missing dependency (responses) needed by `datastore_test.py` and broken-down really long lines in tox.ini for better readability.

zaneselvans commented on tox.ini:
```diff
@@ -113,7 +114,30 @@ commands =
 commands =
-    pytest {posargs} --disable-warnings --log-cli-level=info --log-cli-format="%(asctime)s [%(levelname)8s] %(name)s:%(lineno)s %(message)s" --log-date-format="%Y-%m-%d %H:%M:%S" --cov-report=term --cov-report=xml --cov-config=.coveragerc --cov={envsitepackagesdir}/pudl test/etl_test.py test/glue_test.py test/fast_output_test.py test/interim_etl_test.py test/zenodo_integration.py
+    pytest \
+      --cov-report=term --cov-report=xml --cov-config=.coveragerc \
+      --cov={envsitepackagesdir}/pudl \
+      src/pudl
+    pytest {posargs} \
+      --disable-warnings --log-cli-level=info \
+      --log-cli-format="%(asctime)s [%(levelname)8s] %(name)s:%(lineno)s %(message)s" \
+      --log-date-format="%Y-%m-%d %H:%M:%S" \
+      --cov-report=term --cov-report=xml --cov-config=.coveragerc \
+      --cov={envsitepackagesdir}/pudl \
+      --cov-append \
+      test/etl_test.py \
+      test/glue_test.py \
+      test/fast_output_test.py \
+      test/interim_etl_test.py \
+      test/zenodo_integration.py
```
omg, why did I not realize we could write it like this before. This is so much nicer.
